### PR TITLE
[cxx-interop] Fix exporting cdecl Swift functions to Obj-C++

### PR DESCRIPTION
**Pull request by Xazax-hun**
We export cdecl function declarations twice: for Objective-C and for C++. When the code is compiled in Objective-C++ both of the declarations are visible to the compiler. The generated header did not compile, because only one of the declarations were noexcept. There are multiple possible ways to fix this issue, one of them would make only C++ declarations visible in Objective-C++ mode. However, for this particular problem I decided to also make the Objective-C functions SWIFT_NOEXCEPT. This approach resolves the inconsistency that broke the code when compiled in Objective-C++ mode. Moreover, Swift guarantees that those cdecl declarations cannot raise errors, so in case we only generate the C declarations and consume them from C++ or Objective-C++, those are the correct declarations.

rdar://129550313

### CODE REVIEW

**Comment by Xazax-hun** [comment]
@swift-ci please smoke test

**Review by egorzhdan — APPROVED** [review]
Agreed on the approach chosen, but let's wait for a review from @beccadax to be certain – I'm not sure if there can be any ABI concerns with making this change to Obj-C interop.

**Review by beccadax — APPROVED** [review]
swift/include/swift/PrintAsClang/ClangMacros.def says:

```cpp
CLANG_MACRO_CXX("SWIFT_NOEXCEPT", , "noexcept", )
```

If I understand this correctly, that means `SWIFT_NOEXCEPT` exists in both C++ and ObjC modes, evaluating to `noexcept` in C++ mode and nothing in ObjC mode.

If that's true, then this change LGTM.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp
Comment (drodriguez) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:1443]: Shouldn't `noexcept` be the first thing after the function declaration and before any possible attributes? Does the test below try to parse the header or just checks the output?

Comment (Xazax-hun) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:1443]: Yeah, the `RUN: %check-interop-cxx-header-in-clang(%t/functions.h)` runline will try to parse the header. I used this order for consistency, because other parts of the codebase was also emitting those attributes first and noexcept later. I think there is no strict place to put these non-standard attributes on the declaration, Clang seems to be forgiving on that front.

Comment (drodriguez) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:1443]: The parsing is not as forgiving on my side. In Xcode 16b2 if I add

```
extern int _Nonnull foo(void) __attribute__((warn_unused_result)) noexcept;
```

to a C++ file, I get:

```
error: expected function body after function declarator
   10 | extern int _Nonnull foo(void) __attribute__((warn_unused_result)) noexcept;
      |                                                                   ^
```

Where does the `noexcept` is added in other parts of the codebase? I just sent #74780, but if there's more places it might not be bad to look at them too. They might need fixing.

Comment (Xazax-hun) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:1443]: I just checked and it looks like the order was an oversight on my part. Too bad the `-fsyntax-only` wasn't good enough to catch this :/ Thanks a lot for the PR, I just approved it. We also need to cherry-pick this to the release branch.

**Pull request merged**
